Article: New recommendation API (#339)
qdrant/landing_page: qdrant-landing/content/articles/new-recommendation-api.md

short_description: Qdrant 1.6 brings recommendations strategies and more flexibility to the Recommendation API.
description: Qdrant 1.6 brings recommendations strategies and more flexibility to the Recommendation API.
preview_dir: /articles_data/new-recommendation-api/preview
social_preview_image: /articles_data/new-recommendation-api/preview/social_preview.jpg
small_preview_image: /articles_data/new-recommendation-api/icon.svg
weight: -80
author: Kacper Łukawski
author_link: https://medium.com/@lukawskikacper
date: 2023-10-19T10:12:00.000Z
---

The most popular use case for vector search engines, such as Qdrant, is Semantic search with a single query vector. Given the 
query, we can vectorize (embed) it and find the closest points in the index. But [Vector Similarity beyond Search](/articles/vector-similarity-beyond-search/) 
does exist, and recommendation systems are a great example. Recommendations might be seen as a multi-aim search, where we want 
to find items close to positive and far from negative examples. This use of vector databases has many applications, including 
recommendation systems for e-commerce, content, or even dating apps.

> Qdrant has provided the [Recommendation API](https://qdrant.tech/documentation/concepts/search/#recommendation-api) 
> for a while, but the previous version had some limitations. The latest release, [Qdrant 1.6](https://github.com/qdrant/qdrant/releases/tag/v1.6.0), 
> gives you some new exciting features, including an extension of the recommendation API, making it more flexible and powerful. 
> This article unveils the internals and shows how they may be used in practice.

Reviewer: ```suggestion
Qdrant has provided the [Recommendation API](https://qdrant.tech/documentation/concepts/search/#recommendation-api) for a while, and with the latest release, [Qdrant 1.6](https://github.com/qdrant/qdrant/releases/tag/v1.6.0), 
we're glad to give you more flexibility and control over the Recommendation API. 
Here, we'll discuss some internals and shows how they may be used in practice.
```

Simplified and more casual tone

> ### Recap of the old recommendations API

Reviewer: ```suggestion
### What Recommendations already has
```

> The old [Recommendation API](https://qdrant.tech/documentation/concepts/search/#recommendation-api) in Qdrant came with some limitations. First of all, it was required to pass vector IDs for 

Reviewer: ```suggestion
The previous [Recommendation API](https://qdrant.tech/documentation/concepts/search/#recommendation-api) in Qdrant came with some limitations. First of all, it was required to pass vector IDs for 
```

both positive and negative example points. If you wanted to use vector embeddings directly, you had to either create a new point 
in a collection or mimic the behaviour of the Recommendation API by using the [Search API](https://qdrant.tech/documentation/concepts/search/#search-api).
Moreover, in the previous releases of Qdrant, you were always asked to provide at least one positive example. This requirement 
was based on the algorithm used to combine multiple samples into a single query vector. It was a simple, yet effective approach. 
However, if the only information you had was that your user dislikes some items, you couldn't use it directly.

> Qdrant 1.6 brings a more flexible API. You can now provide both IDs and vectors of positive and negative examples. You can even 
> combine them within a single request. That makes the new implementation backward compatible, so you can easily upgrade an existing
> Qdrant instance without any changes in your code. And the default behaviour of the API is still the same as before. However, we 
> extended the API, so **you can now choose the strategy of how to find the recommended points**.

Reviewer: ```suggestion
Qdrant 1.6 brings a more flexible API. You can now provide both IDs and vectors of positive and negative examples. You can even combine them within a single request. That makes the new implementation backward compatible, so you can easily upgrade an existing Qdrant instance without any changes in your code. And the default behaviour of the API is still the same as before. However, we extended the API, so **you can now choose the strategy of how to find the recommended points**.
```

Fix uneven sentence breaking? This is one place but there might be more which isn't visible here

might want to consider other criteria, such as the ticket price, or avoid some specific locations due to certain restrictions. 
That means, there are various strategies for choosing the best match, which is also true in the case of vector recommendations. 
We can use different approaches to determine the path of traversing the HNSW graph by changing how we calculate the score of a 
candidate point during traversal. The default behaviour is based on pure distance, but Qdrant 1.6 exposes two strategies for the 
recommendation API.

### Average vector

The default strategy, called `average_vector` is the previous one, based on the average of positive and negative examples. It 
simplifies the recommendations process and converts it into a single vector search. It supports both point IDs and vectors as 
parameters. For example, you can get recommendations based on past interactions with existing points combined with query vector 
embedding. Internally, that mechanism is based on the averages of positive and negative examples and was calculated with the 
following formula:

> ```
> average_vector = avg(positive_vectors) + ( avg(positive_vectors) - avg(negative_vectors) )
> ```

Reviewer: ```suggestion
$$
\text{average vector} = \text{avg}(\text{positive vectors}) + \left( \text{avg}(\text{positive vectors}) - \text{avg}(\text{negative vectors}) \right)
$$
```

Should we use [MathJax](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions) here? 

More important: This is not an average vector directly? This is a centroid of positive vectors and then some distance addition to it based on the diff? It'd be helpful to understand what this diff addition is doing for us

Author: No, it's not. We have that described already, but in a different article: https://qdrant.tech/articles/vector-similarity-beyond-search/#vector-features-recommendations

samples and uses a slightly more sophisticated algorithm under the hood.

The best score is chosen at every step of HNSW graph traversal. We separately calculate the distance between a traversed point 
and every positive and negative example. In the case of the best score strategy, **there is no single query vector anymore, but a 
bunch of positive and negative queries**. As a result, for each sample in the query, we have a set of distances, one for each 
sample. In the next step, we simply take the best scores for positives and negatives, creating two separate values. Best scores 
are just the closest distances of a query to positives and negatives. The idea is: **if a point is closer to any negative than to 
any positive example, we do not want it**. We penalize being close to the negatives, so instead of using the similarity value 
directly, we check if it’s closer to positives or negatives. The following formula is used to calculate the score of a traversed 
potential point:

```rust
if best_positive_score > best_negative_score {
    score = best_positive_score
} else {
    score = -(best_negative_score * best_negative_score)
}

> ```

Reviewer: Would you want to use Python instead of Rust here? Though the logic is fairly simple to read

Author: It's just a copy from our codebase, as far as I know. I'd keep it as it is.

If the point is closer to the negatives, we penalize it by taking the negative squared value of the best negative score. For a 
closer negative, the score of the candidate point will always be lower or equal to zero, making the chances of choosing that point 
significantly lower. However, if the best negative score is higher than the best positive score, we still prefer those that are 

> further away from the negatives. That procedure effectively **pulls the traversal procedure away from the negative examples**.

Reviewer: Clarifying why the square of the negative score is used (as opposed to other forms of penalization) could add a little more colour and context perhaps?

Author: I guess that's just one of the options available, and I don't want to discuss the details that much.
